### PR TITLE
libflux: update flux msg route functions

### PR DESCRIPTION
**Pull request by chu11**
I'll just post what is in the primary commit message here:

```
Problem: The functions flux_msg_get_route_first() and
flux_msg_get_route_last() return route IDs via an allocated string
that must be freed by the user.  This API made sense previously
when route IDs were stored internally in a zframe.  However, now
that the route IDs are stored in decoded data structures, we could
return the route IDs directly.

Solution: Return internally stored route IDs via a const char *
instead of an allocated string.  Adjust all callers accordingly.
use 'const char *' instead of 'char *' in several places where
more appropriate.
```

If you're wondering why don't I put this change in one of the other libflux refactoring PRs, its b/c `flux-sched` needs updating too, so figure a separate PR would be best.

I had some further design thoughts too which I'll post in #3617.

### CODE REVIEW

**Comment by garlick** [comment]
If we're going to change the API anyway, maybe we ought to update it to look more like other similar interfaces in flux:
- Use a prefix of `flux_msg_route_`
- Accessors return the route or NULL rather than int success and double pointer paramter
- make some simple functions return void

So maybe
```c
void flux_msg_route_enable (flux_msg_t *msg); 
void flux_msg_route_clear (flux_msg_t *msg);
int flux_msg_route_push (flux_msg_t *msg, const char *id);
char *flux_msg_route_pop (flux_msg_t *msg);
const char *flux_msg_route_first (flux_msg_t *msg);
const char *flux_msg_route_last (flux_msg_t *msg);
int flux_msg_route_count (flux_msg_t *msg);

// wondering if this is needed externally?
char *flux_msg_route_string (const flux_msg_t *msg);

// wondering if this is worth it, given it is easier now to access routes and compare?
bool flux_msg_route_match_first (const flux_msg_t *msg1, const flux_msg_t *msg2);

```

**Comment by chu11** [comment]
@garlick sounds like a good idea. I assume your removal of `const flux_msg_t *` in some of the above is accidental?  Or intentional to perhaps include a `next/prev` in the future?  Will have to see how the current users of the API may/may not break.

**Comment by garlick** [comment]
Sorry, yup that was an accident.

**Comment by chu11** [comment]
> const char *flux_msg_route_first (flux_msg_t *msg);
> const char *flux_msg_route_last (flux_msg_t *msg);

I tried to make this conversion in an earlier prototyping attempt, but reverted it.  Unlike some other functions we have, a return of a NULL id is sort of ok, indicating there are no routes.  So NULL can't be used as an indicator of error.  So I kept the style the same.

What's the opinion on the NULL means no routes or error style?   When I made the conversion earlier I did add something like:

```
if (!(route = flux_msg_get_route_first (msg))) {
    errno = EPROTO;
    goto error;
}
```

in a few places as a consequence.  I felt it was too much of a change, but that's just me of course.

**Comment by chu11** [comment]
> void flux_msg_route_enable (flux_msg_t *msg); 
> void flux_msg_route_clear (flux_msg_t *msg);

given the rest of the msg API doesn't return void, I think it'd be better to return int on error? e.g. these funcs call `flux_msg_set_flags()` internally.

**Comment by garlick** [comment]
> given the rest of the msg API doesn't return void, I think it'd be better to return int on error? e.g. these funcs call flux_msg_set_flags() internally.

It seems like "invalid flag" (which can't happen) or "null message" (which by the time you're setting flags, you probably know you have a good message) are the main possible errors.  Since it is painfully repetitious to check for errors from these functions, it seems to me like these could just not crash on a NULL msg and be OK.

I felt the same way about the route accessors.  Return NULL for "no route" when there is no route, or if there is an error.  By the time you're accessing the routes, you probably know you have a good message.

I might be missing some subtle error case though.  (But if there is one we should decide if it's worth it).

**Comment by chu11** [comment]
> // wondering if this is needed externally?
> char *flux_msg_route_string (const flux_msg_t *msg);

The only user outside of `libflux/message` is `broker/ping`.  I don't think we can remove this function until we have a way for the user to iterate the route frames.

```
// wondering if this is worth it, given it is easier now to access routes and compare?
bool flux_msg_route_match_first (const flux_msg_t *msg1, const flux_msg_t *msg2);
```

w/ the newer `libflux/disconnect` library, the use has plummeted to 3 locations (one of which is disconnect).

Ehhh ... it's sort of borderline.  My lazy immediate thought is to keep the function, otherwise we'd have to do:

```
const char *s1 = flux_msg_route_first (msg1);
const char *s2 = flux_msg_route_first (msg2);
if (!s1 || !s2)
   return false;
if (strcmp (s1, s2))
  return false;
return true;
```

in 3 places :-|

**Comment by chu11** [comment]
re-pushed with many cleanups per discussion above and a few extra pre-cleanups (borderline could peel them off into a new PR, but it's not quite there I think).

Note that the commits:

```
    libflux: return routes via const in msg API
    libflux/message: returns const routes directly
```

could be squashed, but I let them be separate for easier review for now.

Since there's no way to know if a return of NULL is an error or empty route, in some cases I set `errno = EPROTO` in the error path when using the new `flux_msg_route_first/last()`.

Returning void for `flux_msg_route_enable()` had an interesting side effect in librouter.  I had to break up a few big if statements.

```
@@ -51,13 +51,17 @@ static void notify_disconnect (struct service *ss, const char *uuid)
 
     if (!(msg = flux_request_encode ("disconnect", NULL))
         || flux_msg_set_noresponse (msg) < 0
-        || flux_msg_enable_route (msg) < 0
-        || flux_msg_set_cred (msg, ss->cred) < 0
-        || flux_msg_push_route (msg, uuid) < 0
-        || flux_requeue (ss->h, msg, FLUX_RQ_TAIL) < 0) {
-        if (ss->verbose)
-            log_msg ("error notifying server of %.5s disconnect", uuid);
-    }
+        || flux_msg_set_cred (msg, ss->cred) < 0)
+        goto error;
+    flux_msg_enable_route (msg);
+    if (flux_msg_push_route (msg, uuid) < 0
+        || flux_requeue (ss->h, msg, FLUX_RQ_TAIL) < 0)
+        goto error;
+    flux_msg_decref (msg);
+    return;
+error:
+    if (ss->verbose)
+        log_msg ("error notifying server of %.5s disconnect", uuid);
     flux_msg_decref (msg);
 }

```

I could have done some trickery to avoid this, but figured I the average person might be confused if I did `|| (flux_msg_route_enable (), false)`.  Even I'd look at that in a month and wonder what the hell I did.

**Comment by garlick** [comment]
Maybe `push_route()` should internally set the route flag?

**Comment by chu11** [comment]
> Maybe push_route() should internally set the route flag?

Doh, i thought about that last night and just forgot to try it.  I'll give it a shot.

**Comment by garlick** [comment]
Were you amenable to the rename suggestion above?  E.g. use `flux_msg_route_` prefix on these functions?

Also would it be OK to have `route_pop()` just return NULL on both "no route" and error?  Again, it's one of those situations where this doesn't get called until ample error checking has already been performed.  This function has a relatively small number of users.

**Comment by chu11** [comment]
> Were you amenable to the rename suggestion above? E.g. use flux_msg_route_ prefix on these functions?

yup, it worked out fine.

> Also would it be OK to have route_pop() just return NULL on both "no route" and error? 

We can, but the semantics of that function are different.  B/c we're popping a route frame, we're destroying the internal copy.  So we sort of always have to `strdup()` and return the route id.  But a significant amount of the time (5/7 current uses), we don't even need the route id.  So we'd be strdup-ing & free-ing for nothing.  And I think a fair amount of those pop-ings are in pretty important paths.

But the inconsistent API style can be annoying.  Perhaps we just need another function, `flux_msg_route_peek()` or `flux_msg_route_top()`?  Or a `flux_msg_route_del_top()`?

**Comment by garlick** [comment]
Did you already push the name changes?  I wasn't seeing them in my copy.

**Comment by chu11** [comment]
> Edit: oh I just saw that "push_route" and "pop_route" wasn't renamed. I see the other functions were renamed.

Doh!  Sorry,  I missed those.  I was probably processing the removal of `get` in those and forgot to the flip the back part.

**Comment by garlick** [comment]
I'm thinking maybe we ought to bump this to the next release and try to tag soon.  That ok @chu11?

**Comment by chu11** [comment]
@garlick that's fine, i had assumed this was going to be for the next release already.  Minimally it requires flux-sched to be changed too.

**Comment by chu11** [comment]
>  Maybe push_route() should internally set the route flag?

Did a prototype of this change.  It's fine, but leads to an obvious question, do we even need `flux_msg_route_enable()` anymore?

I imagine it was needed in the old `zmsg_t` days b/c you needed to know if you had a delimiter on the list.  But with the new decoded implementation, is it needed?  Presumably a user should just be able to push routes to a message without any "enabling" of it.

**Comment by garlick** [comment]
I think you do need to add a delimiter with no routes before sending a request from a ROUTER socket.  The socket then pushes the identity on before the message is sent to a peer.  So there are cases where you want the delimiter but not routes.

Reading through the ZMQ_ROUTER section of [zmq_socket](https://linux.die.net/man/3/zmq_socket) I'm now questioning whether the delimiter is necessary!  It may just be a hold over from ZMQ_REQ / ZMQ_REP which nobody uses.

**Comment by chu11** [comment]
Re-pushed.  This PR is starting to get big, so may want to start peeling off soon.

I ended up dis-liking having `flux_msg_push_route()` enable routes if they aren't enabled already.  After prototyping, it leads to some inconsistency in how the code is called.  i.e. sometimes call `enable_route()`, sometimes don't.  My anal-retentive side disliked it all.

Against my better judgement, I instead did the C trick w/ a comment to avoid breaking up those big if statements.

```
    /* flux_msg_enable_route() returns void.  To avoid creating an
     * extra branch, call with C trick to avoid breaking up single if
     * statement into multiple.
     */
    if (flux_msg_get_type (msg, &type) < 0
        || type != FLUX_MSGTYPE_REQUEST
        || (flux_msg_enable_route (msg), false)
```

I know I might get overruled on this :-)  Or maybe there's another way I'm not thinking about.

Per discussion above, added `flux_msg_route_delete()`, had `flux_msg_route_pop()` return `char *`.

Fixed up the other function renames that I missed before.

**Comment by chu11** [comment]
removed WIP as I think we're approaching something reviewable & worth merging.  peeled off cleanup commits into #3762 to shrink this PR a bit.

**Comment by chu11** [comment]
was thinking about this last night.  There is always a strong desire to not break the API and be forced to break the `flux-sched` build, so we want to put all API changing things into one PR.

But it's also hard to move forward on #3617 where there isn't the greatest clarity on how it will eventually end up looking.

Should we consider this PR at this point a merge-able candidate?  Accepting potential API changing risk down the road?  Here's the summary of the changes:

` `flux_msg_enable_route()` and `flux_msg_clear_route()` return void
- `flux_msg_pop_route()` returns `char *` instead of returning id via in-and-out parameter
- `flux_msg_get_route_first()` and `flux_msg_get_route_last()` return `const char *` instead of returning id via in-and-out-parameter
- add `flux_msg_delete_route()` to handle the `pop-route` but don't care about the id case (i.e. don't allocate/free memory)
- rename all functions and consistently use the prefix `flux_msg_route`

Keep in mind https://github.com/flux-framework/rfc/pull/295, which would change / obsolete a number of the above changes.  Although I commented in #3617 that I think the proposed change to RFC3 is probably not worthwhile.

**Comment by garlick** [comment]
Yeah, seems like the fixed length route ID didn't really pan out.  OK let's move ahead on this then.

I'll do a review here and then withdraw the RFC 3 PR.

File: src/common/libflux/message.c
Comment (garlick) [inline src/common/libflux/message.c]: Problem: returns NULL on both error (with errno set) and if `list_top()` returns NULL (without errno set).

I think we talked about this but it didn't immediately pop out when I scanned the discussion in this pr - should this function and  other route accessors dispense with trying to set errno altogether?

**Review by garlick — COMMENTED** [review]
Some minor stuff plus some questions inline that might require further discussion.

File: src/common/libflux/message.h
Comment (chu11) [reply src/common/libflux/message.h]: sounds like a good idea. Only subtlety is that `disable` may not imply that all present routes may be cleared.  However, I cannot imagine any use case were a user would call `disable` and want the currently set routes to still be there if they later called `enable`.  As long as its clear what's going on in the comments it should be ok.

Comment (chu11) [reply src/common/libflux/message.h]: To my surprise, what I said above about not imagining a use case ... well we do that in `librouter`.

```
static flux_msg_t *request_copy_clear_routes (const flux_msg_t *msg)                                                                        
{                                                                                                                                           
    flux_msg_t *cpy;                                                                                                                        
    if (!(cpy = flux_msg_copy (msg, true)))                                                                                                 
        return NULL;                                                                                                                        
    flux_msg_clear_route (cpy);                                                                                                             
    flux_msg_enable_route (cpy);                                                                                                            
    return cpy;                                                                                                                             
}
```

I'm now thinking we should have a `enable`, `disable` (which will also clear), and `clear`.  I think that makes a lot more sense.

Comment (garlick) [reply src/common/libflux/message.h]: Meh, it's one internal user.  It would be fine to call disable + enable there.  We can add a comment if needed.

Comment (chu11) [reply src/common/libflux/message.h]: For the time being decided to have enable, disable, and clear.  There is atleast one other location where it's sort of weird to call `disable`.  `clear` makes a lot more sense.

File: src/common/libflux/message.h
Comment (chu11) [reply src/common/libflux/message.h]: good catch, i had missed that.  I think the only thing about using `route_last()` and `route_delete_last()` is there an be some trickiness with the error handling.  But I doubt it's enough trouble to say no to doing this and remove the memory allocation/free forever.

File: src/common/libflux/message.c
Comment (chu11) [reply src/common/libflux/message.c]: i actually went back and forth on the setting errno.  I feel we shouldn't set errno, but I ultimately decided to set errno simply b/c I saw other parts of flux-core do it when the API style was similar.  We're probably not super consistent about how we do this.

Since you mention it, lets go back to not setting errno, since I think that makes more sense.

**Comment by chu11** [comment]
re-pushed per comments above

- `flux_msg_route_delete()` renamed to `flux_msg_route_delete_last()`

- added `flux_msg_route_disable()` which does that `flux_msg_route_clear()` did before.  `flux_msg_route_clear()` now only does a clear but not a disable.

- remove `flux_msg_route_pop()`.  Please note how I adjusted things in `librouter`.

- `flux_msg_route_first()` and `flux_msg_route_last()` no longer set errno.

- some cleanups as a result of above (e.g. `flux_msg_route_match_first()` now uses `flux_msg_route_first()`)

- re-ordered some commits, since the order sort of made no sense given above changes and the commit messages had to be changed.

**Review by garlick — APPROVED** [review]
Alright this is looking good.  I'm fine if this goes in when you are ready.  Does the sched PR need any updating?

**Comment by chu11** [comment]
@garlick Yup, the `flux-sched` PR is all ready here: https://github.com/flux-framework/flux-sched/pull/852

I can't set MWP here b/c of the `flux-sched` breakage.  Apologies, I can't quite remember the workaround so one can manually merge this.

**Comment by grondo** [comment]
To manually merge we'll have to remove branch protections, merge, then re-establish branch protections.

I can help with that when you are ready.

**Comment by chu11** [comment]
@grondo, thanks  Just re-based, so once it passes we can do a manual merge, then i'll kick the flux-sched side to re-do its tests.

**Comment by chu11** [comment]
code coverage builder timed out, restarted builders

**Comment by chu11** [comment]
since this wasn't yet merged, I snuck one more cleanup in.  Should be reviewed before merging.  The cleanup is in commit:

```
commit 0b19862b51d920a7984b5fae56d45e904f75afe5
Author: Albert Chu <chu11@llnl.gov>
Date:   Fri Jul 9 11:31:41 2021 -0700

    libflux: do not reuse zframes when sending msg
```

**Comment by chu11** [comment]
sorry, i snuck in one more cleanup

```
commit 516f7d5918dc3e8d25fe936b605aa24d46fdb401
Author: Albert Chu <chu11@llnl.gov>
Date:   Sat Jul 10 10:59:38 2021 -0700

    libflux: fix const declaration style
    
    Problem: A pointer was declared "uint8_t const *", where as it is
    far more common in flux to use "const uint8_t *".
```

**Comment by chu11** [comment]
found yet another minor thing to cleanup, so I went ahead and peeled all those cleanups onto #3771, so that should go in first.

**Comment by chu11** [comment]
rebased.  None of the "snuck in" commits are here anymore, just the ones that @garlick reviewed.

**Comment by grondo** [comment]
Ok, I'm going to remove branch protections, merge, and then reset branch protections.

**Pull request merged**
